### ISSUE TITLE
Best practices for polyfills?

### ISSUE THREAD

**Issue by veltman** [rollup/rollup#1459]
Is there any particular guidance on the best way to include a polyfill directly (rather than through Babel or something)?

For example, if I were using https://www.npmjs.com/package/whatwg-fetch, presumably I should only import it in one place. Will just importing it at the top of the bundle entry script reliably put it first in line for evaluation?

Relatedly - is the `context` option only bundlewide? Some polyfills (like the above) rely on `this` as `window`, but it seems like a blunt instrument to set it for the entire bundle and potentially produce side effects elsewhere (maybe this is a question for `rollup-plugin-commonjs` instead).

**Comment by bfred-it** [comment]
I'll answer the first question: just make sure it's the first import. [Try it yourself in the repl](https://rollupjs.org/repl?version=0.43.0&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMGElMjBmcm9tJTIwJy4lMkZhLmpzJyUzQiU1Q25pbXBvcnQlMjBiJTIwZnJvbSUyMCcuJTJGYi5qcyclM0IlNUNuY29uc29sZS5sb2coYSUyQ2IpJTIyJTdEJTJDJTdCJTIybmFtZSUyMiUzQSUyMmEuanMlMjIlMkMlMjJjb2RlJTIyJTNBJTIyY29uc29sZS5sb2coJ2EnKSUzQiU1Q25leHBvcnQlMjBkZWZhdWx0JTIwZnVuY3Rpb24lMjAoKSU3QiU3RCUyMiU3RCUyQyU3QiUyMm5hbWUlMjIlM0ElMjJiLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmNvbnNvbGUubG9nKCdiJyklM0IlNUNuZXhwb3J0JTIwZGVmYXVsdCUyMGZ1bmN0aW9uJTIwKCklN0IlN0QlMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyaWlmZSUyMiUyQyUyMm1vZHVsZU5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlMkMlMjJtb2R1bGVJZCUyMiUzQSUyMiUyMiU3RCUyQyUyMmV4YW1wbGUlMjIlM0FudWxsJTdE)

As you can see, `import a from './a.js';` is replaced by the content of `a.js`, keeping the order.

**Comment by Rich-Harris** [comment]
> Will just importing it at the top of the bundle entry script reliably put it first in line for evaluation?

Yep — unless that import somehow imported one of your source files, which would never be the case with a polyfill.

> Relatedly - is the `context` option only bundlewide?

There's also a `moduleContext` option — I *think* this should work:

```js
// rollup.config.js
export default {
  ...,
  moduleContext: {
    'whatwg-fetch': 'window'
  }
};
```

If not, then this might...

```js
// rollup.config.js
export default {
  ...,
  moduleContext: {
    [require.resolve('whatwg-fetch')]: 'window'
  }
};
```

**Comment by veltman** [comment]
Thanks! Looks like the second (fully resolved path) is the trick.

**Comment by piotr-cz** [comment]
Unfortunately either module context `whatwg-fetch` nor `[require.resolve('whatwg-fetch')]` didn't work for me (`(!) `this` has been rewritten to `undefined``)

I'm using rollup v0.56.4 and whatwg-fetch v2.0.3

**Comment by devert** [comment]
For folks who encounter this error, and are not able to get the above solutions to work, what fixed it for me was...

```
moduleContext: {
   'node_modules/whatwg-fetch/fetch.js': 'window',
},
```

**Comment by shirakaba** [comment]
A note for anyone ending up here, wondering how to polyfill specifically the `fetch` implementation from the `whatwg-fetch` library:

It appears, possibly as of [this commit](https://github.com/github/fetch/commit/89d29f76ebd4e8baa8dab77d6cd50015492a46f6) on 21st June 2020, which I believe landed in `whatwg-fetch@3.3.0`, that their fetch implementation no longer uses `this` at the top level and only uses it within classes (which is perfectly sensible). So we don't need to add a `moduleContext` entry for `whatwg-fetch` these days (this is the case at the time of writing, for `whatwg-fetch@3.6.2`).

To globally polyfill `whatwg-fetch@3.6.2`, all you need to do is import it for side-effects in your entrypoint file (for me, that's `src/main.js`). Make sure to import it before any module that itself uses `fetch`, or equally imports another module that uses `fetch`:

```js
import 'whatwg-fetch';
```
